File: ios/Classes/AccountManagerHostApiImpl.swift
import Foundation

/// Implements [AccountManagerHostApi] using iOS Keychain, UserDefaults-backed
/// AccountStore, and BGTaskScheduler-backed BackgroundSyncManager.
class AccountManagerHostApiImpl: AccountManagerHostApi {

    private let keychainManager = KeychainManager.shared
    private let accountStore = AccountStore.shared

    // MARK: - Account Operations

    func addAccount(account: AccountData, password: String, completion: @escaping (Result<Bool, Error>) -> Void) {
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                try self.keychainManager.storeCredentials(
                    username: account.username,
                    password: password,
                    accountType: account.accountType
                )
                try self.accountStore.saveAccount(account)
                DispatchQueue.main.async { completion(.success(true)) }
            } catch {
                DispatchQueue.main.async { completion(.failure(error)) }
            }
        }
    }

    func getAccounts(accountType: String, completion: @escaping (Result<[AccountData], Error>) -> Void) {
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let accounts = try self.accountStore.getAccounts(ofType: accountType)
                DispatchQueue.main.async { completion(.success(accounts)) }
            } catch {
                DispatchQueue.main.async { completion(.failure(error)) }
            }
        }
    }

    func getAccount(username: String, accountType: String, completion: @escaping (Result<AccountData?, Error>) -> Void) {
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let account = try self.accountStore.getAccount(username: username, accountType: accountType)
                DispatchQueue.main.async { completion(.success(account)) }
            } catch {
                DispatchQueue.main.async { completion(.failure(error)) }
            }
        }
    }

    func updateAccount(account: AccountData, completion: @escaping (Result<Bool, Error>) -> Void) {
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                try self.accountStore.updateAccount(account)
                DispatchQueue.main.async { completion(.success(true)) }
            } catch {
                DispatchQueue.main.async { completion(.failure(error)) }
            }
        }
    }

    func removeAccount(account: AccountData, completion: @escaping (Result<Bool, Error>) -> Void) {
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                try self.keychainManager.deleteCredentials(
                    username: account.username,
                    accountType: account.accountType
                )
                try self.accountStore.removeAccount(
                    username: account.username,
                    accountType: account.accountType
                )
                DispatchQueue.main.async { completion(.success(true)) }
            } catch {
                DispatchQueue.main.async { completion(.failure(error)) }
            }
        }
    }

    func accountExists(username: String, accountType: String, completion: @escaping (Result<Bool, Error>) -> Void) {
        let exists = accountStore.accountExists(username: username, accountType: accountType)
        completion(.success(exists))
    }

    // MARK: - Credential Operations

    func updateCredentials(account: AccountData, newPassword: String, completion: @escaping (Result<Bool, Error>) -> Void) {
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                try self.keychainManager.storeCredentials(
                    username: account.username,
                    password: newPassword,
                    accountType: account.accountType
                )
                DispatchQueue.main.async { completion(.success(true)) }
            } catch {
                DispatchQueue.main.async { completion(.failure(error)) }
            }
        }
    }

    func validateCredentials(username: String, password: String, accountType: String, completion: @escaping (Result<Bool, Error>) -> Void) {
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let stored = try self.keychainManager.retrieveCredentials(
                    username: username,
                    accountType: accountType
                )
                DispatchQueue.main.async { completion(.success(stored == password)) }
            } catch {
                DispatchQueue.main.async { completion(.success(false)) }
            }
        }
    }

    func clearCredentials(account: AccountData, completion: @escaping (Result<Bool, Error>) -> Void) {
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                try self.keychainManager.deleteCredentials(
                    username: account.username,
                    accountType: account.accountType
                )
                DispatchQueue.main.async { completion(.success(true)) }
            } catch {
                DispatchQueue.main.async { completion(.failure(error)) }
            }
        }
    }

    // MARK: - Auth Token Operations

    func getAuthToken(account: AccountData, tokenType: String, completion: @escaping (Result<AuthTokenResult, Error>) -> Void) {
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let token = try self.keychainManager.retrieveAuthToken(
                    username: account.username,
                    accountType: account.accountType,
                    tokenType: tokenType
                )
                let result = AuthTokenResult(
                    token: token,
                    errorCode: token == nil ? -1 : nil,
                    errorMessage: token == nil ? "No token available" : nil,
                    requiresUserInteraction: token == nil ? true : nil
                )
                DispatchQueue.main.async { completion(.success(result)) }
            } catch {
                let result = AuthTokenResult(
                    token: nil,
                    errorCode: -1,
                    errorMessage: error.localizedDescription,
                    requiresUserInteraction: nil
                )
                DispatchQueue.main.async { completion(.success(result)) }
            }
        }
    }

    func setAuthToken(account: AccountData, tokenType: String, token: String, completion: @escaping (Result<Bool, Error>) -> Void) {
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                try self.keychainManager.storeAuthToken(
                    username: account.username,
                    accountType: account.accountType,
                    tokenType: tokenType,
                    token: token
                )
                DispatchQueue.main.async { completion(.success(true)) }
            } catch {
                DispatchQueue.main.async { completion(.failure(error)) }
            }
        }
    }

    func invalidateAuthToken(accountType: String, token: String, completion: @escaping (Result<Bool, Error>) -> Void) {
        // iOS Keychain doesn't provide a direct way to look up by token value;
        // invalidation by type is handled via invalidateAllTokens.
        completion(.success(true))
    }

    func invalidateAllTokens(account: AccountData, tokenType: String, completion: @escaping (Result<Bool, Error>) -> Void) {
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                try self.keychainManager.deleteAllTokens(
                    username: account.username,
                    accountType: account.accountType,
                    tokenType: tokenType
                )
                DispatchQueue.main.async { completion(.success(true)) }
            } catch {
                DispatchQueue.main.async { completion(.failure(error)) }
            }
        }
    }

    func getAvailableTokenTypes(account: AccountData, completion: @escaping (Result<[String], Error>) -> Void) {
        completion(.success([]))
    }


    // MARK: - Platform-Specific

    func openAccountSettings(completion: @escaping (Result<Bool, Error>) -> Void) {
        // iOS doesn't have a centralised account settings page
        completion(.success(false))
    }

    func isConfigured(completion: @escaping (Result<Bool, Error>) -> Void) {
        completion(.success(true))
    }

    func getPlatformCapabilities(completion: @escaping (Result<[String: Bool], Error>) -> Void) {
        let caps: [String: Bool] = [
            "systemAccountSettings": false,
            "backgroundSync": true,
            "keychainStorage": true,
            "cloudKitSync": false,
            "pushNotificationSync": true,
            "biometricAuth": true,
        ]
        completion(.success(caps))
    }
}

File: android/src/main/kotlin/com/lkrjangid/account_manager/AccountManagerHostApiImpl.kt
package com.lkrjangid.account_manager

import android.accounts.Account
import android.accounts.AccountManager
import android.content.ContentResolver
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.provider.Settings
import com.lkrjangid.account_manager.sync.SyncManager
import com.lkrjangid.account_manager.utils.toAccountData
import com.lkrjangid.account_manager.utils.toAndroidAccount
import com.lkrjangid.account_manager.utils.toBundle
import kotlinx.coroutines.*

/** Implements [AccountManagerHostApi] using Android's AccountManager system service. */
class AccountManagerHostApiImpl(
    private val context: Context,
    private val syncManager: SyncManager,
) : AccountManagerHostApi {

    private val accountManager: AccountManager = AccountManager.get(context)
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    fun dispose() {
        scope.cancel()
    }

    // -------------------------------------------------------------------------
    // Account Operations
    // -------------------------------------------------------------------------

    override fun addAccount(
        account: AccountData,
        password: String,
        callback: (Result<Boolean>) -> Unit,
    ) {
        scope.launch {
            try {
                val androidAccount = account.toAndroidAccount()
                val success = accountManager.addAccountExplicitly(
                    androidAccount,
                    password,
                    account.userData?.toBundle(),
                )
                if (success) {
                    account.displayName?.let {
                        accountManager.setUserData(androidAccount, "displayName", it)
                    }
                    val authority = "${account.accountType}.provider"
                    ContentResolver.setSyncAutomatically(androidAccount, authority, true)
                }
                withContext(Dispatchers.Main) { callback(Result.success(success)) }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) { callback(Result.failure(e)) }
            }
        }
    }

    override fun getAccounts(
        accountType: String,
        callback: (Result<List<AccountData>>) -> Unit,
    ) {
        scope.launch {
            try {
                val accounts = accountManager.getAccountsByType(accountType)
                val list = accounts.map { it.toAccountData(accountManager) }
                withContext(Dispatchers.Main) { callback(Result.success(list)) }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) { callback(Result.failure(e)) }
            }
        }
    }

    override fun getAccount(
        username: String,
        accountType: String,
        callback: (Result<AccountData?>) -> Unit,
    ) {
        scope.launch {
            try {
                val found = accountManager.getAccountsByType(accountType)
                    .firstOrNull { it.name == username }
                    ?.toAccountData(accountManager)
                withContext(Dispatchers.Main) { callback(Result.success(found)) }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) { callback(Result.failure(e)) }
            }
        }
    }

    override fun updateAccount(
        account: AccountData,
        callback: (Result<Boolean>) -> Unit,
    ) {
        scope.launch {
            try {
                val androidAccount = account.toAndroidAccount()
                account.displayName?.let {
                    accountManager.setUserData(androidAccount, "displayName", it)
                }
                account.userData?.forEach { (k, v) ->
                    if (k != null) accountManager.setUserData(androidAccount, k, v)
                }
                withContext(Dispatchers.Main) { callback(Result.success(true)) }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) { callback(Result.failure(e)) }
            }
        }
    }

    override fun removeAccount(
        account: AccountData,
        callback: (Result<Boolean>) -> Unit,
    ) {
        scope.launch {
            try {
                val androidAccount = account.toAndroidAccount()
                @Suppress("DEPRECATION")
                val success = accountManager.removeAccountExplicitly(androidAccount)
                withContext(Dispatchers.Main) { callback(Result.success(success)) }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) { callback(Result.failure(e)) }
            }
        }
    }

    override fun accountExists(
        username: String,
        accountType: String,
        callback: (Result<Boolean>) -> Unit,
    ) {
        scope.launch {
            try {
                val exists = accountManager.getAccountsByType(accountType)
                    .any { it.name == username }
                withContext(Dispatchers.Main) { callback(Result.success(exists)) }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) { callback(Result.failure(e)) }
            }
        }
    }

    // -------------------------------------------------------------------------
    // Credential Operations
    // -------------------------------------------------------------------------

    override fun updateCredentials(
        account: AccountData,
        newPassword: String,
        callback: (Result<Boolean>) -> Unit,
    ) {
        scope.launch {
            try {
                val androidAccount = account.toAndroidAccount()
                accountManager.setPassword(androidAccount, newPassword)
                withContext(Dispatchers.Main) { callback(Result.success(true)) }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) { callback(Result.failure(e)) }
            }
        }
    }

    override fun validateCredentials(
        username: String,
        password: String,
        accountType: String,
        callback: (Result<Boolean>) -> Unit,
    ) {
        scope.launch {
            try {
                val androidAccount = Account(username, accountType)
                val stored = accountManager.getPassword(androidAccount)
                withContext(Dispatchers.Main) {
                    callback(Result.success(stored != null && stored == password))
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) { callback(Result.failure(e)) }
            }
        }
    }

    override fun clearCredentials(
        account: AccountData,
        callback: (Result<Boolean>) -> Unit,
    ) {
        scope.launch {
            try {
                val androidAccount = account.toAndroidAccount()
                accountManager.setPassword(androidAccount, null)
                withContext(Dispatchers.Main) { callback(Result.success(true)) }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) { callback(Result.failure(e)) }
            }
        }
    }

    // -------------------------------------------------------------------------
    // Auth Token Operations
    // -------------------------------------------------------------------------

    override fun getAuthToken(
        account: AccountData,
        tokenType: String,
        callback: (Result<AuthTokenResult>) -> Unit,
    ) {
        scope.launch {
            try {
                val androidAccount = account.toAndroidAccount()
                val cached = accountManager.peekAuthToken(androidAccount, tokenType)
                if (cached != null) {
                    withContext(Dispatchers.Main) {
                        callback(Result.success(AuthTokenResult(token = cached)))
                    }
                    return@launch
                }
                val future = accountManager.getAuthToken(
                    androidAccount, tokenType, null, false, null, null
                )
                val bundle = future.result
                val token = bundle.getString(AccountManager.KEY_AUTHTOKEN)
                withContext(Dispatchers.Main) {
                    if (token != null) {
                        callback(Result.success(AuthTokenResult(token = token)))
                    } else {
                        callback(
                            Result.success(
                                AuthTokenResult(
                                    errorCode = -1L,
                                    errorMessage = "No token available",
                                    requiresUserInteraction =
                                        bundle.containsKey(AccountManager.KEY_INTENT),
                                )
                            )
                        )
                    }
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    callback(
                        Result.success(
                            AuthTokenResult(errorCode = -1L, errorMessage = e.message)
                        )
                    )
                }
            }
        }
    }

    override fun setAuthToken(
        account: AccountData,
        tokenType: String,
        token: String,
        callback: (Result<Boolean>) -> Unit,
    ) {
        scope.launch {
            try {
                val androidAccount = account.toAndroidAccount()
                accountManager.setAuthToken(androidAccount, tokenType, token)
                withContext(Dispatchers.Main) { callback(Result.success(true)) }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) { callback(Result.failure(e)) }
            }
        }
    }

    override fun invalidateAuthToken(
        accountType: String,
        token: String,
        callback: (Result<Boolean>) -> Unit,
    ) {
        scope.launch {
            try {
                accountManager.invalidateAuthToken(accountType, token)
                withContext(Dispatchers.Main) { callback(Result.success(true)) }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) { callback(Result.failure(e)) }
            }
        }
    }

    override fun invalidateAllTokens(
        account: AccountData,
        tokenType: String,
        callback: (Result<Boolean>) -> Unit,
    ) {
        scope.launch {
            try {
                val androidAccount = account.toAndroidAccount()
                val token = accountManager.peekAuthToken(androidAccount, tokenType)
                if (token != null) {
                    accountManager.invalidateAuthToken(account.accountType, token)
                }
                withContext(Dispatchers.Main) { callback(Result.success(true)) }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) { callback(Result.failure(e)) }
            }
        }
    }

    override fun getAvailableTokenTypes(
        account: AccountData,
        callback: (Result<List<String>>) -> Unit,
    ) {
        scope.launch {
            try {
                // Android doesn't expose available token types natively; return empty list.
                withContext(Dispatchers.Main) { callback(Result.success(emptyList())) }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) { callback(Result.failure(e)) }
            }
        }
    }

    // -------------------------------------------------------------------------
    // Sync Operations
    // -------------------------------------------------------------------------

    override fun syncNow(
        account: AccountData,
        expedited: Boolean,
        callback: (Result<SyncResultData>) -> Unit,
    ) {
        scope.launch {
            try {
                val result = syncManager.requestSync(account.username, account.accountType, expedited)
                withContext(Dispatchers.Main) { callback(Result.success(result)) }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    callback(
                        Result.success(
                            SyncResultData(success = false, errorCode = -1L, errorMessage = e.message)
                        )
                    )
                }
            }
        }
    }

    override fun setSyncAutomatically(
        account: AccountData,
        enabled: Boolean,
        callback: (Result<Boolean>) -> Unit,
    ) {
        scope.launch {
            try {
                val androidAccount = account.toAndroidAccount()
                val authority = "${account.accountType}.provider"
                ContentResolver.setSyncAutomatically(androidAccount, authority, enabled)
                withContext(Dispatchers.Main) { callback(Result.success(true)) }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) { callback(Result.failure(e)) }
            }
        }
    }

    override fun isSyncAutomatically(
        account: AccountData,
        callback: (Result<Boolean>) -> Unit,
    ) {
        scope.launch {
            try {
                val androidAccount = account.toAndroidAccount()
                val authority = "${account.accountType}.provider"
                val result = ContentResolver.getSyncAutomatically(androidAccount, authority)
                withContext(Dispatchers.Main) { callback(Result.success(result)) }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) { callback(Result.failure(e)) }
            }
        }
    }

    override fun addPeriodicSync(
        account: AccountData,
        config: PeriodicSyncConfig,
        callback: (Result<Boolean>) -> Unit,
    ) {
        scope.launch {
            try {
                val androidAccount = account.toAndroidAccount()
                val authority = "${account.accountType}.provider"
                ContentResolver.removePeriodicSync(androidAccount, authority, Bundle.EMPTY)
                ContentResolver.addPeriodicSync(
                    androidAccount,
                    authority,
                    config.extras?.toBundle() ?: Bundle.EMPTY,
                    config.intervalSeconds.toLong(),
                )
                withContext(Dispatchers.Main) { callback(Result.success(true)) }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) { callback(Result.failure(e)) }
            }
        }
    }

    override fun removePeriodicSync(
        account: AccountData,
        callback: (Result<Boolean>) -> Unit,
    ) {
        scope.launch {
            try {
                val androidAccount = account.toAndroidAccount()
                val authority = "${account.accountType}.provider"
                ContentResolver.removePeriodicSync(androidAccount, authority, Bundle.EMPTY)
                withContext(Dispatchers.Main) { callback(Result.success(true)) }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) { callback(Result.failure(e)) }
            }
        }
    }

    override fun getSyncStatus(
        account: AccountData,
        callback: (Result<SyncStatus>) -> Unit,
    ) {
        scope.launch {
            try {
                val status = when {
                    syncManager.isSyncActive(account.username, account.accountType) -> SyncStatus.ACTIVE
                    syncManager.isSyncPending(account.username, account.accountType) -> SyncStatus.PENDING
                    else -> SyncStatus.IDLE
                }
                withContext(Dispatchers.Main) { callback(Result.success(status)) }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) { callback(Result.failure(e)) }
            }
        }
    }

    override fun cancelSync(
        account: AccountData,
        callback: (Result<Boolean>) -> Unit,
    ) {
        scope.launch {
            try {
                syncManager.cancelSync(account.username, account.accountType)
                withContext(Dispatchers.Main) { callback(Result.success(true)) }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) { callback(Result.failure(e)) }
            }
        }
    }

    // -------------------------------------------------------------------------
    // Platform-Specific
    // -------------------------------------------------------------------------

    override fun openAccountSettings(callback: (Result<Boolean>) -> Unit) {
        try {
            val intent = Intent(Settings.ACTION_SYNC_SETTINGS).apply {
                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            }
            context.startActivity(intent)
            callback(Result.success(true))
        } catch (e: Exception) {
            callback(Result.failure(e))
        }
    }

    override fun isConfigured(callback: (Result<Boolean>) -> Unit) {
        callback(Result.success(true))
    }

    override fun getPlatformCapabilities(callback: (Result<Map<String, Boolean>>) -> Unit) {
        val caps = mapOf(
            "systemAccountSettings" to true,
            "backgroundSync" to true,
            "keychainStorage" to false,
            "cloudKitSync" to false,
            "pushNotificationSync" to false,
            "biometricAuth" to false,
        )
        callback(Result.success(caps))
    }
}
